echopype/calibrate/cal_params.py

```diff
@@ -1,155 +1,249 @@
-        out_dict[p] = user_cal_dict.get(p, beam[p])
-
-    # Params from the Vendor_specific group
-    params_from_vend = ["sa_correction", "gain_correction"]
-    for p in params_from_vend:
-        # substitute if p not in user input
-        out_dict[p] = user_cal_dict.get(p, get_vend_cal_params_power(beam=beam, vend=vend, param=p))
+    # Make channel a sorted list
+    if not isinstance(channel, (list, xr.DataArray)):
+        raise ValueError("'channel' has to be a list or an xr.DataArray")
+    else:
+        if isinstance(channel, xr.DataArray):
+            channel_sorted = sorted(channel.data)
+        else:
+            channel_sorted = sorted(channel)
+
+    # Screen parameters: only retain those defined in CAL_PARAMS
+    #  -- transform params in scalar or list to xr.DataArray
+    #  -- directly pass through those that are xr.DataArray
+    out_dict = dict.fromkeys(CAL_PARAMS[sonar_type])
+    for p_name, p_val in user_dict.items():
+        if p_name in out_dict:
+            # if p_val an xr.DataArray, check existence and coordinates
+            if isinstance(p_val, xr.DataArray):
+                # if 'channel' is a coordinate, it has to match that of the data
+                if "channel" in p_val.coords:
+                    if not (sorted(p_val.coords["channel"].data) == channel_sorted):
+                        raise ValueError(
+                            f"The 'channel' coordinate of {p_name} has to match "
+                            "that of the data to be calibrated"
+                        )
+                elif "cal_channel_id" in p_val.coords and "cal_frequency" in p_val.coords:
+                    if not (sorted(p_val.coords["cal_channel_id"].data) == channel_sorted):
+                        raise ValueError(
+                            f"The 'cal_channel_id' coordinate of {p_name} has to match "
+                            "that of the data to be calibrated"
+                        )
+                else:
+                    raise ValueError(
+                        f"{p_name} has to either have 'channel' as a coorindate "
+                        "or have both 'cal_channel_id' and 'cal_frequency' as coordinates"
+                    )
+                out_dict[p_name] = p_val
+
+            # If p_val a scalar or list, make it xr.DataArray
+            elif isinstance(p_val, (int, float, list)):
+                # check for list dimension happens within param2da()
+                out_dict[p_name] = param2da(p_val, channel)
+
+            # p_val has to be one of int, float, xr.DataArray
+            else:
+                raise ValueError(f"{p_name} has to be a scalar, list, or an xr.DataArray")
+
+            # TODO: Consider pre-sort the param xr.DataArray?
 
     return out_dict
 
 
-def get_vend_filter_EK80(
-    vend: xr.Dataset, channel_id: str, filter_name: str, param_type: str
-) -> Union[np.ndarray, int]:
+def _get_interp_da(
+    da_param: Union[None, xr.DataArray],
```

Comment:
```suggestion
    da_param: Optional[xr.DataArray] = None,
```

Reply:
I see, I guess this is a private function, so only internal user would use this. In that case, we can just keep it the way it is.